classes/privacy/provider.php

```diff
@@ -24,19 +24,165 @@
+        $context = $userlist->get_context();
+        if (!$context instanceof context_block) {
+            return;
+        }
+        $sql = "SELECT UNIQUE(userid) FROM {block_dedication}";
+        $userlist->add_from_sql('userid', $sql, []);
+    }
+
+    /**
+     * Delete multiple users within a single context.
+     *
+     * @param approved_userlist $userlist The approved context and user information to delete information for.
+     */
+    public static function delete_data_for_users(approved_userlist $userlist) {
+        global $DB;
+        $context = $userlist->get_context();
+        if (!$context instanceof context_block) {
+            return;
+        }
+
+        $userids = $userlist->get_userids();
+        list($insql, $inparams) = $DB->get_in_or_equal($userids, SQL_PARAMS_NAMED);
+
+        $DB->delete_records_select('block_dedication', "userid $insql", $inparams);
+    }
+
+    /**
+     * Get the list of contexts that contain user information for the specified user.
+     *
+     * In the case of attendance, that is any attendance where a student has had their
+     * attendance taken or has taken attendance for someone else.
+     *
+     * @param int $userid The user to search.
+     * @return contextlist $contextlist The contextlist containing the list of contexts used in this plugin.
+     */
+    public static function get_contexts_for_userid(int $userid): contextlist {
+        return (new contextlist)->add_from_sql(
+            "SELECT cx.id
+            FROM {block_dedication} bd
+            JOIN {context} cx ON cx.contextlevel = :blocklevel AND cx.instanceid = bd.id
+            WHERE bd.userid = :userid",
+            [
+                'blocklevel' => CONTEXT_BLOCK,
+                'userid' => $userid,
+            ]
+        );
+    }
+
+    /**
+     * Delete all personal data for all users in the specified context.
+     *
+     * @param context $context Context to delete data from.
+     */
+    public static function delete_data_for_all_users_in_context(\context $context) {
+        if (!$context instanceof context_block) {
+            return;
+        }
+
+        // The only way to delete data for the html block is to delete the block instance itself.
+        if ($blockinstance = static::get_instance_from_context($context)) {
+            blocks_delete_instance($blockinstance);
+        }
+    }
+
+    /**
+     * Delete all user data for the specified user, in the specified contexts.
+     *
+     * @param approved_contextlist $contextlist The approved contextlist to delete information for.
+     */
+    public static function delete_data_for_user(approved_contextlist $contextlist) {
+        global $DB;
+        $userid = (int) $contextlist->get_user()->id;
+        $params = ['userid' => $userid];
+        $contextids = $contextlist->get_contextids();
+        list($insql, $inparams) = $DB->get_in_or_equal($contextids, SQL_PARAMS_NAMED);
```

Comment:
I'd do 2 things here...

If  ($context instanceof \context_user) {

then delete all data for that user in the dedication table.

if ($context instance of \context_course) {
delete all data for that user in this course.

otherwise just return.